a52 add url to create ownerships as a batch

khelwood opened the pull request.

khelwood committed:
a52 add url to create ownerships as a batch

emrojo commented on app/controllers/ownerships_controller.rb:
```diff
@@ -44,6 +44,58 @@ def destroy
+    modelidhash = {}
+    # I don't know what the expected behaviour is if the same model id is passed twice,
+    #  so abort if a duplicate is found.
+    params.require(:ownership).each do |item|
```
In Rails normally this parameters processing/validation goes into its own method. This allows us to reuse this method in case we want to add new actions that needs the same parameter processing.  You can call it something like 'batch_parameters' and defined it private.

About the parameter validation, there's possibly a way of doing this validation on the content of the array in the ownership parameters without needing to iterate over them. Take a look at the example in:
https://github.com/rails/strong_parameters#nested-parameters

Unfortunately the treatment of duplicates that you are doing cannot be done using the way is done in the documentation. Maybe you can perform it after the validation step.

emrojo commented on app/controllers/ownerships_controller.rb:
```diff
@@ -44,6 +44,58 @@ def destroy
+          status: 400,
+          message: "Repeated model id: #{modelid.inspect}"
+        }
+        return render json: error, status: :bad_request
```
If you separate parameter processing in a different method don't move this logic into it, instead you can keep the condition for this repeated model error inside this action of the controller (as it is now).

emrojo commented on app/controllers/ownerships_controller.rb:
```diff
@@ -44,6 +44,58 @@ def destroy
+    modelids = modelidhash.keys
+
+    Ownership.transaction do
+
```
All this transaction should be refactored inside the model (Ownership class). Anything that is related with CRUD operations should be hidden inside the model. A good approach is to keep the controller as simple as possible, keeping the code for the different renders depending on the conditions of the inputs.

emrojo reviewed (commented):
Some suggestions for refactoring the code.

khelwood committed:
a52 suggested refactors

Separate validation from deduping and more into a separate method,
using permit to check the array contents.
Moved the transaction and its contents to a method in the Ownership
model class.

khelwood committed:
a52 added batch test to ownerships_controller_test.rb

emrojo reviewed (approved):
Merging
